### DISCUSSION TITLE
When accessing the DAG page through the K3S proxy, an error of "404 page not found" is displayed

### DISCUSSION THREAD

**Discussion by microhuang** [apache/airflow#36398] [General] [answered]
### Official Helm Chart version

1.11.0 (latest released)

### Apache Airflow version

2.7.1

### Kubernetes Version

v1.26.4 +k3s1

### Helm Chart configuration

 No configuration has been modified！

### What happened

When accessing the DAG page through the K3S proxy, an error of "404 page not found" is displayed

![图片](https://github.com/apache/airflow/assets/3072926/6637ca30-e8d0-40d8-923b-72a5b1fca1e5)

### How to reproduce

setup:

Install the ranger: https://ranchermanager.docs.rancher.com/v2.7/pages-for-subheaders/rancher-on-a-single-node-with-docker

Install the airflow on local cluster(a k3s cluster): https://airflow.apache.org/docs/helm-chart/1.11.0/index.html

Open: https://127.0.0.1:30443/api/v1/namespaces/airflow/services/http:airflow-webserver:8080/proxy/dags/dataset_consumes_1/grid

![图片](https://github.com/apache/airflow/assets/3072926/edc61dac-3a29-4614-873d-c8efe8c6002f)

![图片](https://github.com/apache/airflow/assets/3072926/6ed86ce4-8c94-439d-88f7-152c38381988)

![图片](https://github.com/apache/airflow/assets/3072926/3eaadebe-5fd7-4d12-80b8-a782a3b6cf36)

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
evidently your poxy is somewhat wrong mapping the URL requests, you likely need to fix it in your configuration somehow. We have no idea what the "ranger" is because it is all in non-english, so I am afraid you will have to find out yourself what to do.

**Reply by microhuang** [reply]
Sorry，"ranger" is a spelling error, it is actually "rancher"。

**Reply by potiuk** [reply]
Still we have no idea what rancher is, but still you should explain in the discussion what you are doing with it and once you do it, what you want to achieve with it and how it works - and while doing it, you will likely find what's wrong by just doing it.

This is not an "airflow isssue" when something you installl does somethign with the URLs and you need to cofigure it to make things work for you. You can continue explaining it in a discussion/troubleshoong and ask for more help - but this is not Airflow issue.

**Reply by microhuang** [reply]
Rancher is a Kubernetes management platform, just like a "dashboard". and I use it to manage k3s clusters. In the k3s cluster, I deployed airflow through "helm upgrade --install airflow apache-airflow/airflow --version 1.11.0 --namespace airflow --create-namespace". When I accessed airflow through the LB service of the k3s cluster, that error occurred.  I guess this is related to "enable_proxy_fix", but I cannot be certain. 


I think the error came from Airflow, not K3s (or Rancher), because by examining the HTML code of the page and its corresponding source code, I found that the same code generated different HTML code. Please see:


https://github.com/apache/airflow/blob/2.7.1/airflow/www/templates/airflow/dag.html => line 62:
`<meta name="grid_url_no_root" content="{{ url_for('Airflow.grid', dag_id=dag.dag_id, num_runs=num_runs_arg, base_date=base_date_arg) }}">`

Expect：
"/api/v1/namespaces/airflow/services/http:airflow-webserver:8080/proxy/dags/dataset_consumes_1/grid"

Result（Error）：
/dags/dataset_consumes_1/grid
![图片](https://github.com/apache/airflow/assets/3072926/fd398185-c1d3-4fee-ae1f-d6adfb36216e)


https://github.com/apache/airflow/blob/2.7.1/airflow/www/templates/airflow/dag.html => line 187:
`<li id="grid-nav"><a href="{{ url_for('Airflow.grid', dag_id=dag.dag_id, num_runs=num_runs_arg, root=root, base_date=base_date_arg) }}">`

Result（Right）：
![图片](https://github.com/apache/airflow/assets/3072926/6e07bb4c-9929-498a-b766-57320127a9a5)

**Reply by potiuk** [reply]
What exactly about "enable proxy fix" ? have you applied it correctly ? You wrote before this:

> No configuration has been modified！

But you know mention `enable_proxy_fix` ? Did you change it? Did you expect to be changed from default `false` valiue? Did you configure everything properly to use the proxy ?

**Reply by potiuk** [reply]
> When I accessed airflow through the LB service of the k3s cluster.

Likely you need to configure it then (possibly via enable_proxy_fix as you mentioned). Airflow is not **guaranteed** to work over not configured proxy services. This is not an "issue" - it's a propertly of how proxies work.

**Reply by microhuang** [reply]
Yes, I haven't made any changes to the configuration because I'm not understand with them. But I wrote a demo with only one line of code： 
`kubectl run hello-alpine --image=alpine:latest --image-pull-policy=IfNotPresent --restart=Never --env="project=k3s" --labels="app=helloworld"  --command -- sh -c 'echo "listing 8080" && while true; do echo -e "HTTP/1.1 200 OK\n\n<!DOCTYPE html>\nhelloworld" | nc -l -k -p 8080 ; done'`
when I accessed it through LB, I got the correct output. LB did not modify my output.

**Reply by microhuang** [reply]
What I mean is, the same code should get the same result. The configuration:

![图片](https://github.com/apache/airflow/assets/3072926/ccf5dbf6-b47b-483e-895b-aa90b4e90c5b)
![图片](https://github.com/apache/airflow/assets/3072926/227a1f64-b890-4e29-9d87-e3487e1b190f)

I also tried to change "enable_proxy_fix", but it didn't work.

**Reply by microhuang** [reply]
> What exactly about "enable proxy fix" ? have you applied it correctly ? You wrote before this:
> 
> > No configuration has been modified！
> 
> But you know mention `enable_proxy_fix` ? Did you change it? Did you expect to be changed from default `false` valiue? Did you configure everything properly to use the proxy ?

In Helm airflow 1.11.0, by default: enable_proxy_fix is Ture:

https://github.com/apache/airflow/blob/helm-chart/1.11.0/chart/values.yaml => line 2267
![图片](https://github.com/apache/airflow/assets/3072926/ddf837f0-acdf-4006-b351-358051d57ffd)

**Reply by potiuk** [reply]
Did you follow the docs? https://airflow.apache.org/docs/apache-airflow/stable/howto/run-behind-proxy.html

**Reply by microhuang** [reply, marked as answer]
I conducted the following experiment:

> A: echo "listing 8080" && while true; do echo -e "HTTP/1.1 200 OK\n**Content-Type: text/html; charset=utf-8**\n\n<"'!'"DOCTYPE html>\n<html><head></head><body><a href=\"/date\">date</a></body></html>" | nc -l -k -p 8080 -q 1; done
> ```Out: <a href="/api/v1/namespaces/airflow273/services/http:test-proxy:8080/proxy/date">date</a>```

> B: echo "listing 8080" && while true; do echo -e "HTTP/1.1 200 OK\n\n<"'!'"DOCTYPE html>\n<html><head></head><body><a href=\"/date\">date</a></body></html>" | nc -l -k -p 8080 -q 1; done
> ```Out: <a href="/date">date</a>```

Confirming that the error came from the rancher, not airflow (or other).

thanks.

**Comment by potiuk** [comment]
Converted it into discussion. If you have any more findings on what is wrong - maybe a good idea to put your findings here for others who migh try similar things (and describing in English what is the purpose of the things you are installing might be helpful too).

**Reply by microhuang** [reply]
This is a recent discovery:



Related configurations:

```
AIRFLOW__ WEBSERVER__ ENABLE_ PROXY_ FIX=True
AIRFLOW__ WEBSERVER__ BASE_ URL=http://localhost:8080/api/v1/namespaces/airflow273/services/http:airflow-webserver:8080/proxy
```


Result of accessing "http://<host-ip>:8080/api/v1/namespaces/airflow273/services/http:airflow-webserver:8080/proxy/home":


> **1. Using nginx proxy -- OK**:

```
listen 8080 default_ Server;

location/api/v1/namespaces/airflow273/services/http:airflow-webserver:8080/proxy/{

Proxy_ Pass http://<pod ip>: 8080;

Proxy_ Set_ Header Host $host;

Proxy_ Redirect off;

Proxy_ HTTP_ Version 1.1;

Proxy_ Set_ Header Upgrade $http_ Upgrade;

Proxy_ Set_ Header Connection "upgrade";

}
```



> **2. Use "kubectl port forward svc/airflow webserver 8080:8080" -- OK**



> **3. Using kubectl proxy -- port=8080 -- ERROR**



View the “https://github.com/k3s-io/kubernetes/blob/v1.26.8-k3s1/staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go Line:44" source file, it can be seen that in "proxy" mode, only a few tags have been correctly processed, not including **redoc** tags and **meta** tags. Should the **base** tag be used in the template to reference the "base_url" value?

![图片](https://github.com/apache/airflow/assets/3072926/338be60e-7107-4ac4-b2cb-a59658a0bc1b)

**Reply by potiuk** [reply]
To be honest - no idea. Seems a bit niche of a case and if someone solves it, cool.
